python/adapters.h

```diff
@@ -19,72 +37,36 @@ py::list rosValueToList(const Embag::PyBindPointerWrapper<Embag::RosValue::Point
   py::list list{};
 
   // TODO: Rather than creating a vector, RosValue should provide an iterator interface
-  for (const auto &value : ros_value->getValues()) {
+  for (const Embag::PyBindPointerWrapper<Embag::RosValue::Pointer, Embag::RosValue> &value : ros_value->getValues()) {
     switch (value->getType()) {
-      case Type::ros_bool: {
-        list.append(value->as<bool>());
-        break;
-      }
-      case Type::int8: {
-        list.append(value->as<int8_t>());
-        break;
-      }
-      case Type::uint8: {
-        list.append(value->as<uint8_t>());
-        break;
-      }
-      case Type::int16: {
-        list.append(value->as<int16_t>());
-        break;
-      }
-      case Type::uint16: {
-        list.append(value->as<uint16_t>());
-        break;
-      }
-      case Type::int32: {
-        list.append(value->as<int32_t>());
-        break;
-      }
-      case Type::uint32: {
-        list.append(value->as<uint32_t>());
-        break;
-      }
-      case Type::int64: {
-        list.append(value->as<int64_t>());
-        break;
-      }
-      case Type::uint64: {
-        list.append(value->as<uint64_t>());
-        break;
-      }
-      case Type::float32: {
-        list.append(value->as<float>());
-        break;
-      }
-      case Type::float64: {
-        list.append(value->as<double>());
-        break;
-      }
-      case Type::string: {
-        list.append(encodeStrLatin1(value->as<std::string>()));
-        break;
-      }
-      case Type::ros_time: {
-        list.append(value->as<Embag::RosValue::ros_time_t>());
-        break;
-      }
+      case Type::ros_bool:
+      case Type::int8:
+      case Type::uint8:
+      case Type::int16:
+      case Type::uint16:
+      case Type::int32:
+      case Type::uint32:
+      case Type::int64:
+      case Type::uint64:
+      case Type::float32:
+      case Type::float64:
+      case Type::string:
+      case Type::ros_time:
       case Type::ros_duration: {
-        list.append(value->as<Embag::RosValue::ros_duration_t>());
+        list.append(castValue(value));
         break;
       }
       case Type::object: {
-        list.append(rosValueToDict(value));
+        list.append(rosValueToDict(value, always_list_primitive_array));
+        break;
+      }
+      case Type::primitive_array: {
+        list.append(primitiveArrayToPyObject(value, always_list_primitive_array));
         break;
       }
       case Type::array:
-      case Type::primitive_array:
       {
-        list.append(rosValueToList(value));
+        list.append(rosValueToList(value, always_list_primitive_array));
```

Comment:
you probably want to remove the `case` statement for these 3 types in `py::object castValue` i.e. line 137-140 ?